File: mobile/lib/data/models/media.dart
import 'dart:convert';
import '../../core/constants.dart';

class MediaItem {
  final int id;
  final String? name;
  final String? type;
  final String? overview;
  final String? tagline;
  final String? premiereDate;
  final String? endDate;
  final String? officialRating;
  final double? communityRating;
  final double? criticRating;
  final int? productionYear;
  final int? runTimeTicks;
  final String? dateCreated;
  final String? dateModified;
  final bool? hasChildren;
  final String? album;
  final String? artist;
  final String? author;
  final String? publisher;
  final String? publisherYear;
  final List<FileInfo> files;
  final List<LinkInfo> links;
  final UserData? userdata;
  final List<AliasInfo> alias;

  MediaItem({
    required this.id,
    this.name,
    this.type,
    this.overview,
    this.tagline,
    this.premiereDate,
    this.endDate,
    this.officialRating,
    this.communityRating,
    this.criticRating,
    this.productionYear,
    this.runTimeTicks,
    this.dateCreated,
    this.dateModified,
    this.hasChildren,
    this.album,
    this.artist,
    this.author,
    this.publisher,
    this.publisherYear,
    this.files = const [],
    this.links = const [],
    this.userdata,
    this.alias = const [],
  });

  String? getPrimaryImageUrl({String? token}) {
    String withToken(String url) {
      if (token == null || token.isEmpty) return url;
      return '$url&token=${Uri.encodeComponent(token)}';
    }

    for (final f in files) {
      if (f.imageType == 'Primary') {
        return withToken('/api/file/data?file_id=${f.id}');
      }
    }
    if (files.isNotEmpty) {
      final firstImage = files.firstWhere(
        (f) => f.type == 'Image',
        orElse: () => files.first,
      );
      return withToken('/api/file/data?file_id=${firstImage.id}');
    }
    return null;
  }

  MediaType get mediaType {
    if (type == null){
      return MediaType.unknown;
    }else{
      return MediaType.values.firstWhere((e) => e.name == type, orElse: () => MediaType.unknown);
    }
  }

  factory MediaItem.fromJson(Map<String, dynamic> json) {
    return MediaItem(
      id: (json['id'] as num).toInt(),
      name: json['name'] as String?,
      type: json['type'] as String?,
      overview: json['overview'] as String?,
      tagline: json['tagline'] as String?,
      premiereDate: json['premiere_date'] as String?,
      endDate: json['end_date'] as String?,
      officialRating: json['official_rating'] as String?,
      communityRating: (json['community_rating'] as num?)?.toDouble(),
      criticRating: (json['critic_rating'] as num?)?.toDouble(),
      productionYear: (json['production_year'] as num?)?.toInt(),
      runTimeTicks: (json['run_time_ticks'] as num?)?.toInt(),
      dateCreated: json['date_created'] as String?,
      dateModified: json['date_modified'] as String?,
      hasChildren: json['has_children'] as bool?,
      album: json['album'] as String?,
      artist: json['artist'] as String?,
      author: json['author'] as String?,
      publisher: json['publisher'] as String?,
      publisherYear: json['publisher_year'] as String?,
      files: (json['files'] as List<dynamic>?)
              ?.map((e) => FileInfo.fromJson(e as Map<String, dynamic>))
              .toList() ??
          [],
      links: (json['links'] as List<dynamic>?)
              ?.map((e) => LinkInfo.fromJson(e as Map<String, dynamic>))
              .toList() ??
          [],
      userdata: json['userdata'] != null
          ? UserData.fromJson(json['userdata'] as Map<String, dynamic>)
          : null,
      alias: (json['alias'] as List<dynamic>?)
              ?.map((e) => AliasInfo.fromJson(e as Map<String, dynamic>))
              .toList() ??
          [],
    );
  }
}

class MediaListResponse {
  final List<MediaItem> items;
  final int total;
  final int limit;
  final int offset;

  MediaListResponse({
    required this.items,
    required this.total,
    required this.limit,
    required this.offset,
  });

  factory MediaListResponse.fromJson(Map<String, dynamic> json) {
    return MediaListResponse(
      items: (json['items'] as List<dynamic>)
          .map((e) => MediaItem.fromJson(e as Map<String, dynamic>))
          .toList(),
      total: (json['total'] as num?)?.toInt() ?? 0,
      limit: (json['limit'] as num?)?.toInt() ?? 50,
      offset: (json['offset'] as num?)?.toInt() ?? 0,
    );
  }
}

class MediaStats {
  final int videoCount;
  final int audioCount;
  final int imageCount;
  final int ebookCount;

  MediaStats({
    required this.videoCount,
    required this.audioCount,
    required this.imageCount,
    required this.ebookCount,
  });

  factory MediaStats.fromJson(Map<String, dynamic> json) {
    return MediaStats(
      videoCount: (json['video_count'] as num?)?.toInt() ?? 0,
      audioCount: (json['audio_count'] as num?)?.toInt() ?? 0,
      imageCount: (json['image_count'] as num?)?.toInt() ?? 0,
      ebookCount: (json['ebook_count'] as num?)?.toInt() ?? 0,
    );
  }
}

class FileInfo {
  final int id;
  final String? name;
  final String? path;
  final String? type;
  final String? imageType;
  final int? imageIndex;
  final int? size;
  final String? etag;
  final Map<String, dynamic>? ffmpeg;

  FileInfo({
    required this.id,
    this.name,
    this.path,
    this.type,
    this.imageType,
    this.imageIndex,
    this.size,
    this.etag,
    this.ffmpeg,
  });

  factory FileInfo.fromJson(Map<String, dynamic> json) {
    return FileInfo(
      id: (json['id'] as num).toInt(),
      name: json['name'] as String?,
      path: json['path'] as String?,
      type: json['type'] as String?,
      imageType: json['image_type'] as String?,
      imageIndex: (json['image_index'] as num?)?.toInt(),
      size: (json['size'] as num?)?.toInt(),
      etag: json['etag'] as String?,
      ffmpeg: _parseFfmpeg(json['ffmpeg']),
    );
  }

  static Map<String, dynamic>? _parseFfmpeg(dynamic value) {
    if (value == null) return null;
    if (value is Map<String, dynamic>) return value;
    if (value is String) {
      try {
        final decoded = jsonDecode(value);
        if (decoded is Map<String, dynamic>) return decoded;
      } catch (_) {}
    }
    return null;
  }
}

class LinkInfo {
  final String? peopleType;
  final String? peopleRole;
  final String? sourceLink;
  final LinkedItem? linkedItem;

  LinkInfo({this.peopleType, this.peopleRole, this.sourceLink, this.linkedItem});

  factory LinkInfo.fromJson(Map<String, dynamic> json) {
    return LinkInfo(
      peopleType: json['people_type'] as String?,
      peopleRole: json['people_role'] as String?,
      sourceLink: json['source_link'] as String?,
      linkedItem: json['linked_item'] != null
          ? LinkedItem.fromJson(json['linked_item'] as Map<String, dynamic>)
          : null,
    );
  }
}

class LinkedItem {
  final int id;
  final String? name;
  final String? type;
  final String? overview;
  final String? tagline;
  final String? premiereDate;
  final String? officialRating;
  final double? communityRating;
  final String? primaryImage;
  final String? sourceLink;

  LinkedItem({
    required this.id,
    this.name,
    this.type,
    this.overview,
    this.tagline,
    this.premiereDate,
    this.officialRating,
    this.communityRating,
    this.primaryImage,
    this.sourceLink,
  });

  factory LinkedItem.fromJson(Map<String, dynamic> json) {
    return LinkedItem(
      id: (json['id'] as num).toInt(),
      name: json['name'] as String?,
      type: json['type'] as String?,
      overview: json['overview'] as String?,
      tagline: json['tagline'] as String?,
      premiereDate: json['premiere_date'] as String?,
      officialRating: json['official_rating'] as String?,
      communityRating: (json['community_rating'] as num?)?.toDouble(),
      primaryImage: json['primary_image'] as String?,
      sourceLink: json['source_link'] as String?,
    );
  }
}

class UserData {
  final int? playbackPositionTicks;
  final double? playbackRate;
  final int? playCount;
  final bool? isPlayed;
  final double? rating;
  final DateTime? lastPlayedDate;
  final DateTime? favoritedAt;
  final bool? favorite;

  UserData({
    this.playbackPositionTicks,
    this.playbackRate,
    this.playCount,
    this.isPlayed,
    this.rating,
    this.lastPlayedDate,
    this.favoritedAt,
    this.favorite,
  });

  bool get isFavorite => favorite ?? (favoritedAt != null);

  String? get lastPlayedDisplay {
    if (lastPlayedDate != null) {
      final diff = DateTime.now().difference(lastPlayedDate!);
      if (diff.inMinutes < 1) return '刚刚';
      if (diff.inMinutes < 60) return '${diff.inMinutes} 分钟前';
      if (diff.inHours < 24) return '${diff.inHours} 小时前';
      if (diff.inDays < 30) return '${diff.inDays} 天前';
      return '${lastPlayedDate!.year}-${lastPlayedDate!.month.toString().padLeft(2, '0')}-${lastPlayedDate!.day.toString().padLeft(2, '0')}';
    }
    if (isPlayed == true) return '已观看';
    return null;
  }

  factory UserData.fromJson(Map<String, dynamic> json) {
    DateTime? lastPlayed;
    final lastPlayedStr = json['last_played_date'] as String?;
    if (lastPlayedStr != null) {
      lastPlayed = DateTime.tryParse(lastPlayedStr);
    }
    DateTime? favoritedAt;
    final favoritedAtStr = json['favorited_at'] as String?;
    if (favoritedAtStr != null) {
      favoritedAt = DateTime.tryParse(favoritedAtStr);
    }
    return UserData(
      playbackPositionTicks: (json['playback_position_ticks'] as num?)?.toInt(),
      playbackRate: (json['playback_rate'] as num?)?.toDouble(),
      playCount: (json['play_count'] as num?)?.toInt(),
      isPlayed: json['is_played'] as bool?,
      rating: (json['rating'] as num?)?.toDouble(),
      lastPlayedDate: lastPlayed,
      favoritedAt: favoritedAt,
      favorite: json['is_favorite'] as bool?,
    );
  }
}

class AliasInfo {
  final String? name;
  final String? source;

  AliasInfo({this.name, this.source});

  factory AliasInfo.fromJson(Map<String, dynamic> json) {
    return AliasInfo(
      name: json['name'] as String?,
      source: json['source'] as String?,
    );
  }
}

class UpdateUserDataRequest {
  final int itemId;
  final double? playbackPosition;
  final double? playbackRate;
  final bool? isFavorite;
  final bool? isPlayed;
  final int? playCount;
  final double? rating;
  final bool clearRating;

  UpdateUserDataRequest({
    required this.itemId,
    this.playbackPosition,
    this.playbackRate,
    this.isFavorite,
    this.isPlayed,
    this.playCount,
    this.rating,
    this.clearRating = false,
  });

  Map<String, dynamic> toJson() => {
        'item_id': itemId,
        if (playbackPosition != null) 'playback_position': playbackPosition,
        if (playbackRate != null) 'playback_rate': playbackRate,
        if (isFavorite != null) 'is_favorite': isFavorite,
        if (isPlayed != null) 'is_played': isPlayed,
        if (playCount != null) 'play_count': playCount,
        if (clearRating) 'rating': null else if (rating != null) 'rating': rating,
      };
}

/// 媒体列表请求参数
///
/// 封装 /api/media/list 的查询参数，用于横向媒体列表和网格页面之间传递筛选上下文。
class MediaListRequest {
  final Set<MediaType>? types;
  final bool favorite;
  final bool hasPlayback;
  final bool hasRating;
  final String? sortBy;
  final String? itemIds;
  final String? linkedItemIds;
  final String? search;

  const MediaListRequest({
    this.types,
    this.favorite = false,
    this.hasPlayback = false,
    this.hasRating = false,
    this.sortBy,
    this.itemIds,
    this.linkedItemIds,
    this.search,
  });

  /// 序列化为 API 查询参数（仅筛选字段，不含分页参数）
  Map<String, dynamic> toQueryParams() => {
        if (types != null && types!.isNotEmpty)
          'types': types!.map((t) => t.name).join(','),
        if (favorite) 'favorite': true,
        if (hasPlayback) 'has_playback': true,
        if (hasRating) 'has_rating': true,
        if (sortBy != null) 'sort_by': sortBy,
        if (itemIds != null) 'item_ids': itemIds,
        if (linkedItemIds != null) 'linked_item_ids': linkedItemIds,
        if (search != null) 'search': search,
      };

  MediaListRequest copyWith({
    Set<MediaType>? types,
    bool? favorite,
    bool? hasPlayback,
    bool? hasRating,
    String? sortBy,
    String? itemIds,
    String? linkedItemIds,
    String? search,
    bool clearTypes = false,
    bool clearItemIds = false,
    bool clearLinkedItemIds = false,
    bool clearSearch = false,
  }) =>
      MediaListRequest(
        types: clearTypes ? null : (types ?? this.types),
        favorite: favorite ?? this.favorite,
        hasPlayback: hasPlayback ?? this.hasPlayback,
        hasRating: hasRating ?? this.hasRating,
        sortBy: sortBy ?? this.sortBy,
        itemIds: clearItemIds ? null : (itemIds ?? this.itemIds),
        linkedItemIds:
            clearLinkedItemIds ? null : (linkedItemIds ?? this.linkedItemIds),
        search: clearSearch ? null : (search ?? this.search),
      );
}

File: backend/app/schemas/media.py
"""媒体数据序列化 Schema"""

from typing import Optional
from pydantic import BaseModel


class LinkedItemSummary(BaseModel):
    """关联媒体摘要信息"""
    id: int
    name: str
    type: Optional[str] = None
    overview: Optional[str] = None
    tagline: Optional[str] = None
    premiere_date: Optional[str] = None
    official_rating: Optional[str] = None
    community_rating: Optional[float] = None
    primary_image: Optional[str] = None
    source_link: Optional[str] = None


class LinkItem(BaseModel):
    """媒体关联项（人物/角色）"""
    people_type: Optional[str] = None
    people_role: Optional[str] = None
    source_link: Optional[str] = None
    linked_item: LinkedItemSummary


class FileInfo(BaseModel):
    """文件信息"""
    id: int
    name: Optional[str] = None
    path: Optional[str] = None
    type: Optional[str] = None
    link_type: Optional[str] = None
    image_type: Optional[str] = None
    image_index: Optional[int] = None
    size: Optional[int] = None
    etag: Optional[str] = None
    ffmpeg: Optional[str] = None


class FileInfoDetail(BaseModel):
    """文件详情（含关联的 MediaItem ID）"""
    id: int
    name: str
    path: str
    type: Optional[str] = None
    link_type: Optional[str] = None
    item_id: int
    image_type: Optional[str] = None
    image_index: Optional[int] = None
    size: int
    etag: Optional[str] = None
    ffmpeg: Optional[str] = None


class AliasItem(BaseModel):
    """别名"""
    name: str
    source: Optional[str] = None


class UserDataInfo(BaseModel):
    """用户播放数据"""
    is_favorite: bool = False
    playback_position_ticks: int = 0
    playback_rate: float = 1.0
    play_count: int = 0
    is_played: bool = False
    rating: Optional[float] = None
    last_played_date: Optional[str] = None
    favorited_at: Optional[str] = None


class MediaItemSummary(BaseModel):
    """媒体项基本信息"""
    id: int
    name: str
    type: Optional[str] = None
    overview: Optional[str] = None
    tagline: Optional[str] = None
    premiere_date: Optional[str] = None
    end_date: Optional[str] = None
    official_rating: Optional[str] = None
    community_rating: Optional[float] = None
    critic_rating: Optional[float] = None
    date_created: Optional[str] = None
    date_modified: Optional[str] = None
    source_id: Optional[str] = None
    source_link: Optional[str] = None
    source_name: Optional[str] = None


class MediaItemResponse(MediaItemSummary):
    """媒体项完整响应（含关联数据）"""
    links: list[LinkItem] = []
    files: list[FileInfo] = []
    userdata: Optional[UserDataInfo] = None
    alias: list[AliasItem] = []
    has_children: bool = False


class MediaListResponse(BaseModel):
    """媒体列表响应"""
    items: list[MediaItemResponse]
    total: int
    limit: int
    offset: int
    next_cursor: Optional[str] = None


class MediaStatsResponse(BaseModel):
    """媒体统计响应"""
    video_count: int = 0
    audio_count: int = 0
    image_count: int = 0
    subtitle_count: int = 0
    movie_count: int = 0
    series_count: int = 0
    episode_count: int = 0
    book_count: int = 0
    source_count: int = 0


# ========== 序列化函数 ==========

def serialize_links(links_result, primary_images_map: dict[int, str] | None = None) -> list[LinkItem]:
    """序列化关联媒体链接"""
    links = []
    for link, linked_item in links_result:
        primary_image = primary_images_map.get(linked_item.Id) if primary_images_map else None
        links.append(LinkItem(
            people_type=link.PeopleType.value if link.PeopleType else None,
            people_role=link.PeopleRole,
            source_link=None,
            linked_item=LinkedItemSummary(
                id=linked_item.Id,
                name=linked_item.Name,
                type=linked_item.Type.value if linked_item.Type else None,
                overview=linked_item.Overview,
                tagline=linked_item.Tagline,
                premiere_date=linked_item.PremiereDate.isoformat() if linked_item.PremiereDate else None,
                official_rating=linked_item.OfficialRating,
                community_rating=linked_item.CommunityRating,
                primary_image=primary_image,
                source_link=None,
            )
        ))
    return links


def serialize_files(files_result) -> list[FileInfo]:
    """序列化文件列表"""
    files = []
    for file, file_link in files_result:
        fields = {
            'id': file.Id,
            'name': file.Name,
            'path': file.Path,
            'type': file.Type.value if file.Type else None,
            'link_type': file_link.LinkType.value if file_link.LinkType else None,
            'image_type': file_link.ImageType.value if file_link.ImageType else None,
            'image_index': file_link.ImageIndex,
            'etag': file.Etag,
            'ffmpeg': file.FFmpeg,
        }
        if file.Size is not None:
            fields['size'] = file.Size
        files.append(FileInfo(**fields))
    return files


def serialize_alias(alias_list) -> list[AliasItem]:
    """序列化别名列表"""
    return [AliasItem(name=a.Name, source=a.Source) for a in alias_list]


def serialize_userdata(ud) -> Optional[UserDataInfo]:
    """序列化用户播放数据"""
    if not ud:
        return None
    last_played = None
    if ud.LastPlayedAt:
        last_played = ud.LastPlayedAt.isoformat()
    favorited_at = None
    if ud.FavoritedAt:
        favorited_at = ud.FavoritedAt.isoformat()
    return UserDataInfo(
        is_favorite=ud.FavoritedAt is not None,
        playback_position_ticks=ud.PlaybackPositionTicks,
        playback_rate=ud.PlaybackRate,
        play_count=ud.PlayCount,
        is_played=ud.IsPlayed,
        rating=ud.Rating,
        last_played_date=last_played,
        favorited_at=favorited_at,
    )


def serialize_item(item) -> MediaItemSummary:
    """序列化媒体项基本信息"""
    return MediaItemSummary(
        id=item.Id,
        name=item.Name,
        type=item.Type.value if item.Type else None,
        overview=item.Overview,
        tagline=item.Tagline,
        premiere_date=item.PremiereDate.isoformat() if item.PremiereDate else None,
        end_date=item.EndDate.isoformat() if item.EndDate else None,
        official_rating=item.OfficialRating,
        community_rating=item.CommunityRating,
        critic_rating=item.CriticRating,
        date_created=item.DateCreated.isoformat() if item.DateCreated else None,
        date_modified=item.DateModified.isoformat() if item.DateModified else None,
        source_id=item.SourceId,
        source_link=item.SourceLink,
        source_name=None,
    )


def serialize_file_info(file, file_link) -> FileInfoDetail:
    """序列化文件详情"""
    return FileInfoDetail(
        id=file.Id,
        name=file.Name,
        path=file.Path,
        type=file.Type.value if file.Type else None,
        link_type=file_link.LinkType.value if file_link.LinkType else None,
        item_id=file_link.ItemId,
        image_type=file_link.ImageType.value if file_link.ImageType else None,
        image_index=file_link.ImageIndex,
        size=file.Size,
        etag=file.Etag,
        ffmpeg=file.FFmpeg,
    )
